source/connection-monitoring-and-pooling/connection-monitoring-and-pooling.rst

```diff
@@ -70,6 +70,8 @@ Connection Pool Monitoring
 ~~~~~~~~~~~~~~~~~~~~~~~~~~
 
 All drivers that implement a connection pool MUST provide an API that allows users to subscribe to events emitted from the pool.
+Conceptually, event emission is instantaneous, i.e., one may talk about the instant an event is emitted,
+and represents the start of an activity of delivering the event to a subscribed user.
```

Comment:
There are two terms: _emission_, _delivery_. The first one is already used in the spec, the second one is not. When we talked, I said that I am concerned that an implementor may interpret _emission_ as _delivery_. Interestingly, this is exactly what happened to you when you wrote

> Whether or not the event emission is instantaneous, slow, async or sync - it makes no matter.

It's the _delivery_ that may be slow/sync/async (and, as we discussed, I mentioned neither of those words), and can't be instantaneous, because it's an activity. This activity has the beginning and the end. The beginning of the _delivery_ activity is _emission_.

Reply:
Okay - but what we discussed was specifying that we start measuring the duration immediately _before_ emitting an event because we wanted consistency in when drivers started timing checkout durations relative to event emissions.  That is not captured in this PR.

Reply:
> what we discussed was specifying that we start measuring the duration immediately before emitting an event because we wanted consistency in when drivers started timing checkout durations relative to event emissions

It may be that each of us was hearing/meaning a different thing when we discussed this. At the moment I do not understand what ambiguity (or a different issue) is in the PR text.

Durations are defined as the time elapsed between emitting event ``X and event `Y`. Emission is defined as the start of the delivery activity; it is instantaneous, i.e., we can talk about event `X` being emitted at instant `tx` (note how the word "instantaneous" is defined in the spec right away, leaving no ambiguity of what it means). Thus, the duration is `ty - tx`.

Could you tell what undesirable implementation you envision that is allowed by the spec as it currently is in this PR, and what wording added to the spec will forbid that implementation? If you will use the word "immediately" or its derivative, could you also define it?<sup>1</sup>

---

<sup>1</sup> It is usually fine to use the word "immediately" colloquially (we did use it when we talked), but it is not necessary the right way to express an idea precisely. "immediately" is a useful word when we talk about ordering relations (e.g., "in the total order of natural numbers 1 is followed immediately by 2"), but is much less useful when it is used in the temporal meaning, because the latter is vague and seems to be suitable only for informal usage.

Reply:
We're saying the same thing, I just didn't understand the phrasing.  I also wouldn't be surprised if others don't have the same precise understandings of emission and delivery.  When I say immediately, I don't really care about the exact moment temporally.  I am okay with the wording, now that it has been clarified but would probably prefer something that is a little less dependent on the exact definitions of phrases and words (of which we don't define in the specification).

Perhaps, to avoid ambiguity and misunderstandings that stem from language, we could just update the pseudocode?

```
connection = Null
record checkout start time
emit ConnectionCheckOutStartedEvent and equivalent log message
try:
  enter WaitQueue
  wait until at top of wait queue
  # Note that in a lock-based implementation of the wait queue would
  # only allow one thread in the following block at a time
  while connection is Null:
    if a connection is available:
      while connection is Null and a connection is available:
        connection = next available connection
        if connection is perished:
          close connection
          connection = Null
    else if totalConnectionCount < maxPoolSize:
      if pendingConnectionCount < maxConnecting:
        connection = create connection
      else:
        # this waiting MUST NOT prevent other threads from checking Connections
        # back in to the pool.
        wait until pendingConnectionCount < maxConnecting or a connection is available
        continue

except pool is "closed":
  calculate checkout duration
  emit ConnectionCheckOutFailedEvent(reason="poolClosed", duration=duration) and equivalent log message
  throw PoolClosedError
except pool is "paused":
  calculate checkout duration
  emit ConnectionCheckOutFailedEvent(reason="connectionError", duration=duration) and equivalent log message
  throw PoolClearedError
except timeout:
  calculate checkout duration
  emit ConnectionCheckOutFailedEvent(reason="timeout", duration=duration) and equivalent log message
  throw WaitQueueTimeoutError
finally:
  # This must be done in all drivers
  leave wait queue

# If the Connection has not been established yet (TCP, TLS,
# handshake, compression, and auth), it must be established
# before it is returned.
# This MUST NOT block other threads from acquiring connections.
if connection state is "pending":
  try:
    establish connection
  except connection establishment error:
    calculate checkout duration
    emit ConnectionCheckOutFailedEvent(reason="connectionError", duration=duration) and equivalent log message
    decrement totalConnectionCount
    throw
  finally:
    decrement pendingConnectionCount
else:
    decrement availableConnectionCount
set connection state to "in use"

# If there is no background thread, the pool MUST ensure that
# there are at least minPoolSize total connections.
do asynchronously:
  while totalConnectionCount < minPoolSize:
    populate the pool with a connection

calculate checkout duration
emit ConnectionCheckedOutEvent(duration=duration) and equivalent log message
return connection
```

Reply:
Done. Updated the pseudocode to reflect the added `duration`s. Some lines are quite long. I'll split them if you think the code will be more readable that way.